### PR TITLE
feat: [PIE-7997]: Need merge api for rerun page based on planExecutionId

### PR DESCRIPTION
**Pull request by shalini-agr**
## Describe your changes
**Issue:** Need merge api for rerun page based on planExecutionId

**Change:** Added an api 'mergeForRerun' in InputSetResourcePMS, which takes planExecutionId as input and merges runtime inputs used in that execution with the pipeline to return merged pipeline yaml.

**Testing:** Tested the behavior in local by hitting the new API
I have a pipeline with stages cs1 and cs2 with runtime input and cs3 without runtime input 

1. without stage identifiers (cs1 and cs2 have runtime inputs)

Screenshot 2023-02-01 at 7 41 58 PM

2. with stage identifiers: ["cs1", "cs3"] (only cs1 has runtime input)

Screenshot 2023-02-01 at 7 42 24 PM

3. with stage identifier: ["cs3"] (which does not have any runtime input)

Screenshot 2023-02-01 at 7 42 42 PM

4. with stage identifier: ["cs4"] (which does not exist in pipeline)

Screenshot 2023-02-01 at 7 42 57 PM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43342)

### CODE REVIEW

File: pipeline-service/service/src/main/java/io/harness/pms/plan/execution/service/PMSExecutionServiceImpl.java [hunk 1]
```diff
@@ -439,6 +439,24 @@ public InputSetYamlWithTemplateDTO getInputSetYamlWithTemplate(String accountId,
...
+            .findByAccountIdAndOrgIdentifierAndProjectIdentifierAndPlanExecutionIdAndPipelineDeletedNot(
+                accountId, orgId, projectId, planExecutionId, !pipelineDeleted);
+    if (pipelineExecutionSummaryEntityOptional.isPresent()) {
+      PipelineExecutionSummaryEntity executionSummaryEntity = pipelineExecutionSummaryEntityOptional.get();
+
+      // InputSet yaml used during execution
+      return executionSummaryEntity.getInputSetYaml();
+    }
+    throw new InvalidRequestException(
+        "Invalid request : Input Set did not exist or pipeline execution has been deleted");
```
Comment (namanvermah) [inline pipeline-service/service/src/main/java/io/harness/pms/plan/execution/service/PMSExecutionServiceImpl.java:457]: Why would this mean that input set is deleted? And which input set for that matter?

File: pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/resources/InputSetResourcePMSImpl.java [hunk 2]
```diff
@@ -275,6 +275,37 @@ public ResponseDTO<MergeInputSetResponseDTOPMS> getMergeInputSetFromPipelineTemp
                                        .build());
   }
 
+  @NGAccessControlCheck(resourceType = "PIPELINE", permission = PipelineRbacPermissions.PIPELINE_VIEW)
+  public ResponseDTO<MergeInputSetResponseDTOPMS> getMergeInputSetForRerun(@NotNull @AccountIdentifier String accountId,
+      @NotNull @OrgIdentifier String orgIdentifier, @NotNull @ProjectIdentifier String projectIdentifier,
+      @NotNull @ResourceIdentifier String pipelineIdentifier, String pipelineBranch,
+
```
Comment (namanvermah) [inline pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/resources/InputSetResourcePMSImpl.java:282]: Empty line here can be removed

**Comment by mohitg0795** [comment]
*[`pipeline-service/contracts/src/main/java/io/harness/pms/ngpipeline/inputset/resources/InputSetResourcePMS.java` line 407 at r1](https://reviewable.io/reviews/harness/harness-core/43342#-NNBKkhb-Yq4qhYTSzGl:-NNBKkhb-Yq4qhYTSzGm:bqb2wyk) ([raw file](https://github.com/harness/harness-core/blob/6d8633f4747ceb217344ff9c87efef321ae5dab6/pipeline-service/contracts/src/main/java/io/harness/pms/ngpipeline/inputset/resources/InputSetResourcePMS.java#L407)):*
> ```Java
> 
>   @POST
>   @Path("mergeForRerun")
> ```

We should use camel casing for paths. Its much clearer. Also, we can check with API squad if there are any other recommendations on the same.

**Comment by mohitg0795** [comment]
*[`pipeline-service/modules/pms-contracts/src/main/java/io/harness/pms/inputset/MergeInputSetForRerunRequestDTO.java` line 38 at r1](https://reviewable.io/reviews/harness/harness-core/43342#-NNBKyeI645Wxh4m-MRg:-NNBKyeI645Wxh4m-MRh:bykzz7p) ([raw file](https://github.com/harness/harness-core/blob/6d8633f4747ceb217344ff9c87efef321ae5dab6/pipeline-service/modules/pms-contracts/src/main/java/io/harness/pms/inputset/MergeInputSetForRerunRequestDTO.java#L38)):*
> ```Java
>       description =
>           "This is a boolean value that indicates if the response must contain the YAML for the merged Pipeline. The default value is False.")
>   boolean withMergedPipelineYaml;
> ```

the variable name seems ambiguous and doesn't clearly represent what does it actually mean. May be something better would be : eg: "getResponseWithMergedPipelineYaml" or "getMergedPipelineYamlInResponse" or similar.

File: pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/helpers/ValidateAndMergeHelper.java [hunk 3]
```diff
@@ -264,6 +269,21 @@ public String mergeInputSetIntoPipeline(String accountId, String orgIdentifier,
     return mergeInputSetIntoPipelineForGivenStages(pipelineYaml, mergedRuntimeInputYaml, false, stageIdentifiers);
   }
 
+  public String mergeInputSetIntoPipelineForRerun(String accountId, String orgIdentifier, String projectIdentifier,
```
Comment (namanvermah) [inline pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/helpers/ValidateAndMergeHelper.java:272]: Please add a unit test for this

**Review by namanvermah — CHANGES_REQUESTED** [review]

**Review by shalini-agr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43342)* status: 9 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/helpers/ValidateAndMergeHelper.java` line 272 at r1](https://reviewable.io/reviews/harness/harness-core/43342#-NNBDQ4p4_X4sI_22gbV-r1-272:-NNB_h0-EFQQR-OO87-D:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/6d8633f4747ceb217344ff9c87efef321ae5dab6/pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/helpers/ValidateAndMergeHelper.java#L272)):*
<details><summary><i>Previously, namanvermah (Naman Verma) wrote…</i></summary><blockquote>

Please add a unit test for this
</blockquote></details>

Done.
___
*[`pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/resources/InputSetResourcePMSImpl.java` line 282 at r1](https://reviewable.io/reviews/harness/harness-core/43342#-NNBDQ4p4_X4sI_22gbW-r1-282:-NNBLkVBA4jCC8QeNkK8:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/6d8633f4747ceb217344ff9c87efef321ae5dab6/pipeline-service/service/src/main/java/io/harness/pms/ngpipeline/inputset/resources/InputSetResourcePMSImpl.java#L282)):*
<details><summary><i>Previously, namanvermah (Naman Verma) wrote…</i></summary><blockquote>

Empty line here can be removed
</blockquote></details>

Done.
___
*[`pipeline-service/service/src/main/java/io/harness/pms/plan/execution/service/PMSExecutionServiceImpl.java` line 457 at r1](https://reviewable.io/reviews/harness/harness-core/43342#-NNBDQ4p4_X4sI_22gbY-r1-457:-NNBMZgL5htDMQ4mrMTh:b-sddf2g) ([raw file](https://github.com/harness/harness-core/blob/6d8633f4747ceb217344ff9c87efef321ae5dab6/pipeline-service/service/src/main/java/io/harness/pms/plan/execution/service/PMSExecutionServiceImpl.java#L457)):*
<details><summary><i>Previously, namanvermah (Naman Verma) wrote…</i></summary><blockquote>

Why would this mean that input set is deleted? And which input set for that matter?
</blockquote></details>

yeah, this would actually only mean that the pipeline execution has been deleted since we didn't send anything related to input set in the query

**Review by shalini-agr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43342)* status: 9 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/pms-contracts/src/main/java/io/harness/pms/inputset/MergeInputSetForRerunRequestDTO.java` line 38 at r1](https://reviewable.io/reviews/harness/harness-core/43342#-NNBKyeI645Wxh4m-MRg:-NNBaCBN99o5j2TGjuA_:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/6d8633f4747ceb217344ff9c87efef321ae5dab6/pipeline-service/modules/pms-contracts/src/main/java/io/harness/pms/inputset/MergeInputSetForRerunRequestDTO.java#L38)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

the variable name seems ambiguous and doesn't clearly represent what does it actually mean. May be something better would be : eg: "getResponseWithMergedPipelineYaml" or "getMergedPipelineYamlInResponse" or similar.
</blockquote></details>

Done.

**Review by shalini-agr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43342)* status: 9 file(s) to review. 1 change requests. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/contracts/src/main/java/io/harness/pms/ngpipeline/inputset/resources/InputSetResourcePMS.java` line 407 at r1](https://reviewable.io/reviews/harness/harness-core/43342#-NNBKkhb-Yq4qhYTSzGl:-NNBeXgSBn5A7hJUkZN9:b-kr7k2l) ([raw file](https://github.com/harness/harness-core/blob/6d8633f4747ceb217344ff9c87efef321ae5dab6/pipeline-service/contracts/src/main/java/io/harness/pms/ngpipeline/inputset/resources/InputSetResourcePMS.java#L407)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

We should use camel casing for paths. Its much clearer. Also, we can check with API squad if there are any other recommendations on the same.
</blockquote></details>

renamed the path of merge-for-rerun

**Review by namanvermah — APPROVED** [review]
> Reviewed 1 of 6 files at r1, 5 of 6 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43342)* status: 9 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by archit-harness — APPROVED** [review]
> Reviewed 3 of 6 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 1 of 2 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43342)* status: 3 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by ankittiwari-harness — APPROVED** [review]
:lgtm:

> Reviewed 3 of 6 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 1 of 2 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43342#-:-NNFdSVk4R9v4oWwrmC4:bnfp4nl)* status: 2. unresolved discussions.

**Review by namanvermah — APPROVED** [review]
> Reviewed 1 of 2 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43342)* status: 2. unresolved discussions.

**Review by mohitg0795 — APPROVED** [review]
> Reviewed 3 of 6 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 1 of 2 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43342)* status: :shipit: complete! ✓

**Pull request merged**